hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenRelationshipReadableKVState.java

```diff
@@ -0,0 +1,175 @@
+        this.tokenRepository = tokenRepository;
+    }
+
+    @Override
+    protected TokenRelation readFromDataSource(@Nonnull EntityIDPair key) {
+        final var tokenId = key.tokenId();
+        final var accountId = key.accountId();
+        if (tokenId == null
+                || accountId == null
+                || AccountID.DEFAULT.equals(accountId)
+                || TokenID.DEFAULT.equals(tokenId)) {
+            return null;
+        }
+
+        final var timestamp = ContractCallContext.get().getTimestamp();
+        // The accountId will always be in the format "shard.realm.num"
+        return findTokenAccount(tokenId, accountId, timestamp)
+                .map(ta -> tokenRelationFromEntity(tokenId, accountId, ta, timestamp))
+                .orElse(null);
+    }
+
+    @Nonnull
+    @Override
+    protected Iterator<EntityIDPair> iterateFromDataSource() {
+        return Collections.emptyIterator();
+    }
+
+    @Override
+    public long size() {
+        return 0;
+    }
+
+    private Optional<TokenAccount> findTokenAccount(
+            final TokenID tokenID, final AccountID accountID, final Optional<Long> timestamp) {
+        AbstractTokenAccount.Id id = new AbstractTokenAccount.Id();
+        id.setTokenId(toEntityId(tokenID).getId());
+        id.setAccountId(toEntityId(accountID).getId());
+        return timestamp
+                .map(t -> tokenAccountRepository.findByIdAndTimestamp(id.getAccountId(), id.getTokenId(), t))
+                .orElseGet(() -> tokenAccountRepository.findById(id));
+    }
+
+    private TokenRelation tokenRelationFromEntity(
+            final TokenID tokenID,
+            final AccountID accountID,
+            final TokenAccount tokenAccount,
+            final Optional<Long> timestamp) {
+        return TokenRelation.newBuilder()
+                .tokenId(tokenID)
+                .accountId(accountID)
+                .balanceSupplier(getBalance(tokenAccount, timestamp))
+                .frozen(tokenAccount.getFreezeStatus() == TokenFreezeStatusEnum.FROZEN)
+                .kycGranted(tokenAccount.getKycStatus() != TokenKycStatusEnum.REVOKED)
+                .automaticAssociation(tokenAccount.getAutomaticAssociation())
+                .build();
+    }
+
+    /**
+     * For the latest block we have the balance directly as a field in the TokenAccount object.
+     * For the historical block we need to execute a query to calculate the historical balance, but
+     * we first need to find the account created timestamp and the token type in order to use the correct repository.
+     * If the account creation is after the passed timestamp - return 0L (the entity was not created).
+     */
+    private Supplier<Long> getBalance(final TokenAccount tokenAccount, final Optional<Long> timestamp) {
+        return Suppliers.memoize(() -> timestamp
+                .map(ts -> findAccount(toAccountId(tokenAccount.getAccountId()), Optional.of(ts))
```

Comment:
Believe you can just pass the existing one.
```suggestion
                .map(ts -> findAccount(toAccountId(tokenAccount.getAccountId()), timestamp)
```

Reply:
After removing the check for account creation timestamp, we don't need to fetch the account anymore, so this is no more applicable.